#!/usr/bin/env python3

# GLWOA-RRT* Global Planner Dynamic Reconfigure
PACKAGE = "glwoa_rrtstar_planner"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()
# default, min, max

# General parameters
gen.add("use_optimizer", bool_t, 0, "Enable the WOA optimizer after RRT* path planning.", True)
gen.add("number_of_tests", int_t, 0, "Number of runs for both RRT* and WOA optimizer. Displays average runtime and path cost. The best path will be used for navigation.", 1, 1, 300)
gen.add("visualize_tree", bool_t, 0, "Display the RRT* tree and paths in RViz (Marker).", False)

# RRT* parameters
gen.add("goal_tolerance", double_t, 0, "Allowable distance from the goal point to be considered as reached", 0.2, 0.01, 5.0)
gen.add("rewiring_radius", double_t, 0, "Rewiring radius for RRT*. A large value gives better optimization, but increases computation.", 0.5, 0.1, 50.0)
gen.add("epsilon", double_t, 0, "Exploration step for RRT* in meters. A higher value gives faster exploration, but may struggle if the environement is complex.", 0.2, 0.01, 10.0)
gen.add("max_num_nodes", int_t, 0, "The maximum number of nodes for RRT* to execute, used to limit the runtime of RRT* in case no solution is found.", 5000, 500, 200000)
# gen.add("min_num_nodes", int_t, 0, "RRT* will continue to refine the path until this number of nodes is reached.", 2500, 500, 15000)

# WOA parameters
gen.add("sampling_radius", double_t, 0, "Used for biased sampling in WOA initialization. It is recommended to be a smaller value than epsilon for smoother paths, otherwise increase num_iterations.", 0.12, 0.01, 10.0)
gen.add("num_iterations", int_t, 0, "Number of iterations for WOA to run", 400, 50, 10000)
gen.add("num_agents", int_t, 0, "Number of WOA agents (population). Each agent is a path with the same length as RRT* initial path", 100, 5, 10000)
gen.add("spiral_shape", double_t, 0, "Parameter defining the spiral in WOA. A value lower than 1 gives better", 0.8, 0.0, 3.0)

exit(gen.generate(PACKAGE, "glwoa_rrtstar_planner", "GlWoaRrtStarPlanner"))